rust/core/src/ffi/types.rs

```diff
@@ -17,4 +17,622 @@
+        }
+    }
+}
+
+impl FFI_AdbcError {
+    pub fn with_driver(driver: *const FFI_AdbcDriver) -> Self {
+        FFI_AdbcError {
+            private_driver: driver,
+            ..Default::default()
+        }
+    }
+}
+
+impl TryFrom<&FFI_AdbcError> for Error {
+    type Error = Error;
+
+    fn try_from(value: &FFI_AdbcError) -> Result<Self, Self::Error> {
+        let message = match value.message.is_null() {
+            true => "<empty>".to_string(),
+            false => {
+                let message = unsafe { CStr::from_ptr(value.message) };
+                let message = message.to_owned();
+                message.into_string()?
+            }
+        };
+
+        let mut error = Error {
+            message,
+            status: Status::Unknown,
+            vendor_code: value.vendor_code,
+            sqlstate: value.sqlstate,
+            details: None,
+        };
+
+        if value.vendor_code == constants::ADBC_ERROR_VENDOR_CODE_PRIVATE_DATA {
+            if let Some(driver) = unsafe { value.private_driver.as_ref() } {
+                let get_detail_count = driver_method!(driver, ErrorGetDetailCount);
+                let get_detail = driver_method!(driver, ErrorGetDetail);
+                let num_details = unsafe { get_detail_count(value) };
+                let details = (0..num_details)
+                    .map(|i| unsafe { get_detail(value, i) })
+                    .filter(|d| !d.key.is_null() && !d.value.is_null())
+                    .map(|d| unsafe {
+                        let key = CStr::from_ptr(d.key).to_string_lossy().to_string();
+                        let value = std::slice::from_raw_parts(d.value, d.value_length);
+                        (key, value.to_vec())
+                    })
+                    .collect();
+                error.details = Some(details);
+            }
+        }
+
+        Ok(error)
+    }
+}
+
+impl TryFrom<FFI_AdbcError> for Error {
+    type Error = Error;
+
+    fn try_from(value: FFI_AdbcError) -> Result<Self, Self::Error> {
+        (&value).try_into()
+    }
+}
+
+// Invariant: keys.len() == values.len()
+pub(crate) struct ErrorPrivateData {
+    pub(crate) keys: Vec<CString>,
+    pub(crate) values: Vec<Vec<u8>>,
+}
+
+impl TryFrom<Error> for FFI_AdbcError {
+    type Error = Error;
+
+    fn try_from(mut value: Error) -> Result<Self, Self::Error> {
+        let message = CString::new(value.message)?;
+
+        let private_data = match value.details.take() {
+            None => null_mut(),
+            Some(details) => {
+                let keys: Result<Vec<CString>, _> = details
+                    .iter()
+                    .map(|(key, _)| CString::new(key.as_str()))
+                    .collect();
+                let values: Vec<Vec<u8>> = details.into_iter().map(|(_, value)| value).collect();
+
+                let private_data = Box::new(ErrorPrivateData {
+                    keys: keys?,
+                    values,
+                });
```

Comment:
Nit:
```suggestion
                let keys: Vec<CString> = details
                    .iter()
                    .map(|(key, _)| CString::new(key.as_str()))
                    .collect()?;
                let values: Vec<Vec<u8>> = details.into_iter().map(|(_, value)| value).collect();

                let private_data = Box::new(ErrorPrivateData {
                    keys,
                    values,
                });
```

Reply:
The idea here is to get the error of the first `CString::new` that fails, we cannot directly collect as `Vec<Cstring>`.

Reply:
Right, I meant to write:
```rust
let keys = details
    .iter()
    .map(|(key, _)| CString::new(key.as_str()))
    .collect::<Result<Vec<_>, _>>()?;
```